File: LeetCode/2055. Plates Between Candles/solution.py
"""
35 / 35 test cases passed.
Runtime: 276 ms
Memory Usage: 42.5 MB
"""
class Solution:
    def platesBetweenCandles(self, s: str, queries: List[List[int]]) -> List[int]:
        n = len(s)
        left, right, psum = [0] * n, [0] * n, [0] * n
        l, r, p = -1, -1, 0
        for i in range(n):
            if s[i] == '|': l = i
            if s[n - 1 - i] == '|': r = n - 1 - i
            if s[i] == '*': p += 1
            left[i], right[n - 1 - i], psum[i] = l, r, p
        ans = []
        for query in queries:
            x, y = right[query[0]], left[query[1]]
            ans.append(0 if x == -1 or y == -1 or x >= y else psum[y] - psum[x])
        return ans

File: LeetCode/2055. Plates Between Candles/solution.cpp
/**
 * 35 / 35 test cases passed.
 * Runtime: 324 ms
 * Memory Usage: 141.6 MB 
 */
class Solution {
public:
    vector<int> platesBetweenCandles(string s, vector<vector<int>>& queries) {
        int n = s.size(); 
        vector<int> left(n), right(n), psum(n);
        for (int i = 0, t = 0, l = -1, r = -1; i < n; ++ i) {
            if (s[i] == '*') ++ t;
            if (s[i] == '|') l = i;
            if (s[n - 1 - i] == '|') r = n - 1 - i;
            psum[i] = t;
            left[i] = l;
            right[n - 1 - i] = r;
        }
        vector<int> ans;
        for (auto &query : queries) {
            int l = right[query[0]], r = left[query[1]];
            ans.emplace_back(l == -1 || r == -1 || l >= r ? 0 : psum[r] - psum[l]);
        }
        return ans;
    }
};
